RoaringBitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java

```diff
@@ -1338,248 +1338,86 @@ public static RoaringBitmap andNot(final RoaringBitmap x1, final RoaringBitmap x
-      if (newS2 >= s2) {
-        orNotHandleRemainingHoles(this, newS2, lastKey, lastSize);
+    int length1 = highLowContainer.size(), length2 = other.highLowContainer.size();
+    int s1 = length1 > 0 ? highLowContainer.getKeyAtIndex(pos1) : maxKey + 1;
+    int s2 = length2 > 0 ? other.highLowContainer.getKeyAtIndex(pos2) : maxKey + 1;
+    int remainder = 0;
+    for (int i = highLowContainer.size - 1;
+         i >= 0 && highLowContainer.keys[i] > maxKey;
+         --i) {
+      ++remainder;
+    }
+    int correction = 0;
+    for (int i = 0; i < other.highLowContainer.size - remainder; ++i) {
+      correction += other.highLowContainer.getContainerAtIndex(i).isFull() ? 1 : 0;
+      if (other.highLowContainer.getKeyAtIndex(i) >= maxKey) {
+        break;
       }
     }
-  }
-
-
-  private static RoaringBitmap doOrNot(
-          final RoaringBitmap rb1, final RoaringBitmap rb2, long rangeEnd) {
-    final RoaringBitmap answer = new RoaringBitmap();
-
-    int pos1 = 0, pos2 = 0;
-    int length1 = rb1.highLowContainer.size();
-    final int length2 = rb2.highLowContainer.size();
-
-    final char lastKey = Util.highbits(rangeEnd - 1);
-    final int lastSize = (Util.lowbits(rangeEnd - 1));
-
-    char s2 = 0;
-    boolean loopedAtleastOnce = (length1 > 0 && length2 > 0
-            && (char) 0 - lastKey <= 0);
-    while (pos1 < length1 && pos2 < length2
-            && s2 - lastKey <= 0) { // s2 <= lastKey
-      final char s1 = rb1.highLowContainer.getKeyAtIndex(pos1);
-      final int containerLast = (s2 == lastKey) ? lastSize : Util.maxLowBitAsInteger();
-
-      if (s1 == s2) {
-        final Container c1 = rb1.highLowContainer.getContainerAtIndex(pos1);
-        if (s2 == rb2.highLowContainer.getKeyAtIndex(pos2)) {
-          final Container c2 = rb2.highLowContainer.getContainerAtIndex(pos2);
-          final Container c = c1.orNot(c2, containerLast + 1);
-          answer.highLowContainer.append(s1, c);
-          pos2++;
-        } else {
-          answer.highLowContainer.append(s1,
-                  Container.rangeOfOnes(0, containerLast + 1));
-        }
-        pos1++;
-        s2++;
-      } else if (s1 - s2 > 0) { 
-        if (s2 == rb2.highLowContainer.getKeyAtIndex(pos2)) {
-          final Container c2 = rb2.highLowContainer.getContainerAtIndex(pos2);
-          Container c = new RunContainer().orNot(c2, containerLast + 1);
-          answer.highLowContainer.append(s2, c);
-          pos2++;
-        } else {
-          answer.highLowContainer.append(s2,
-                  Container.rangeOfOnes(0, containerLast + 1));
-        }
-        s2++;
-      } else { 
-        throw new IllegalStateException("This is a bug. Please report to github");
-      }
+    // it's almost certain that the bitmap will grow, so make a conservative overestimate,
+    // this avoids temporary allocation, and can trim afterwards
+    int maxSize = Math.min(maxKey + 1 + remainder - correction + highLowContainer.size, 0x10000);
```

Comment:
@lemire I think this estimate is guaranteed to provide a large enough array. Can you convince me otherwise?

Reply:
It's hard to read in the diff, `correction` is the number of containers on the RHS which are full. `remainder` is the number of containers on the LHS beyond the range end.

Reply:
@lemire if you can convince yourself this is correct - I believe it is - please merge.

Reply:
Yep.